src/Http/Authentication.Core/src/AuthenticationService.cs

```diff
@@ -78,7 +78,8 @@ public AuthenticationService(IAuthenticationSchemeProvider schemes, IAuthenticat
             }
 
             var result = await handler.AuthenticateAsync();
-            if (result != null && result.Succeeded)
+
+            if (result.Succeeded)
```

Comment:
Isn't that a breaking change? If a handler compiled without nullable references enabled (e.g because it was compiled for ASP.NET Core 2.1/3.1) returns null, won't that cause a nullref exception here?

Reply:
We could keep this part as a precaution

Reply:
The compiler complains if you check for null. I'd much rather we prefer the earlier version of explicitly stating that this could return null have have behavior where do not expect null values but still guard against them.

Reply:
The compiler warning can be overridden, right?

AuthenticateAsync is not supposed to return null so that's appropriate to enforce in the API. It's also an extensibility point so @kevinchalet's right that we can't stop people from returning null. Keeping one extra check here seems fine.

Reply:
Why not making `IAuthenticationHandler.AuthenticateAsync()` return a "nullable" `AuthenticateResult` and assimilate `null` results as `AuthenticationResult.NoResult()` in `AuthenticationService.AuthenticateAsync()` so that it never returns `null`?

It's still an observable behavior change, but it's likely better than a nullref.

Reply:
I'd like to communicate to IAuthenticationHandler authors not to return null, but converting null to NoResult sounds like a good idea too.

Reply:
> but converting null to NoResult sounds like a good idea too.

Let's do that. It's the least breaking

Reply:
Done